Graduate JobPodFailurePolicy to stable

mimowo opened the pull request:
#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature
/kind api-change
/sig apps

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Tracking issue: https://github.com/kubernetes/enhancements/issues/3329

#### Special notes for your reviewer:

The related code PRs for the graduation:
- https://github.com/kubernetes/kubernetes/pull/125461
- https://github.com/kubernetes/kubernetes/pull/125442
- https://github.com/kubernetes/kubernetes/pull/125482

The PRs are split for ease of reviewing as the changes are independent and may require different expertise / reviewers.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Graduate JobPodFailurePolicy to GA and lock
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures
```

mimowo commented:
/remove-sig node
/remove-area kubelet
After the split of PodDisruptionConditions to https://github.com/kubernetes/kubernetes/pull/125461

mimowo commented:
/test pull-kubernetes-node-e2e-containerd 
it is ticketed already: https://github.com/kubernetes/kubernetes/issues/123852

mimowo commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features
restest, seems infra issues, there is no build log even

fedebongio commented:
/remove-sig api-machinery

dejanzele commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -2593,19 +2588,18 @@ func TestSyncJobWithJobPodFailurePolicy(t *testing.T) {
-		wantStatusSucceeded           int32
-		wantStatusTerminating         *int32
+		enablePodDisruptionConditions      bool
+		TestSyncJobWithJobPodFailurePolicy bool
```
This field doesn't seem to be used anywhere

mimowo replied:
Right, not sure it got there, but removed now :)

dejanzele commented:
/lgtm

dejanzele commented:
/lgtm

bart0sh commented:
/triage accepted

soltysh commented:
/priority important-longterm

soltysh reviewed (approved):
/lgtm
/approve

atiratree commented:
lgtm, also I presume the code gating changes will be done after this PR and https://github.com/kubernetes/kubernetes/pull/125461 are merged, right?

alculquicondor reviewed (commented):
/lgtm
/approve

mimowo commented:
> lgtm, also I presume the code gating changes will be done after this PR and #125461 are merged, right?

Actually, I was thinking we merge the PR as is (combining the obsolete test removal, and feature gate update) as the feature gate is a "test" that we didn't forget to update a test. This approach was taken in the past for example: https://github.com/kubernetes/kubernetes/pull/123575/, or https://github.com/kubernetes/kubernetes/pull/121302/.

mimowo commented:
/cc @liggitt @deads2k

mimowo commented:
Rebased after https://github.com/kubernetes/kubernetes/pull/125461 got merged.

deads2k commented:
/lgtm
/approve

mimowo commented:
`pull-kubernetes-typecheck` and `pull-kubernetes-verify-lint` complained about the extra field `jobPodFailurePolicy` which was used in a new test cases added in https://github.com/kubernetes/kubernetes/pull/125985. I rebased and removed the field use. Please renew lgtm.

mimowo commented:
/test pull-kubernetes-integration 
overloaded infra

alculquicondor commented:
/lgtm

mimowo commented:
One more rebase after https://github.com/kubernetes/kubernetes/pull/126030

mimowo commented:
/retest

mimowo committed:
JobPodFailurePolicy to GA

# Conflicts:
#	pkg/controller/job/job_controller_test.go

mimowo commented:
One more rebase after https://github.com/kubernetes/kubernetes/pull/125510. The conflicts were only inside job_controller_test.go, due to new feature gates fields added in this PR.

deads2k commented:
/lgtm
